extensions/cornerstone-3d/src/services/ViewportService/Cornerstone3DViewportService.ts

```diff
@@ -249,17 +283,147 @@ class Cornerstone3DViewportService implements IViewportService {
     });
   }
 
+  async _setVolumeViewport(
+    viewport: Types.IVolumeViewport,
+    viewportData: VolumeData,
+    viewportInfo: ViewportInfo
+  ): Promise<void> {
+    // TODO: We need to overhaul the way data sources work so requests can be made
+    // async. I think we should follow the image loader pattern which is async and
+    // has a cache behind it.
+    // The problem is that to set this volume, we need the metadata, but the request is
+    // already in-flight, and the promise is not cached, so we have no way to wait for
+    // it and know when it has fully arrived.
+    // loadStudyMetadata(StudyInstanceUID) => Promise([instances for study])
+    // loadSeriesMetadata(StudyInstanceUID, SeriesInstanceUID) => Promise([instances for series])
+    // If you call loadStudyMetadata and it's not in the DicomMetadataStore cache, it should fire
+    // a request through the data source?
+    // (This call may or may not create sub-requests for series metadata)
+    //
+    // If you call loadSeriesMetadata and a request is ongoing for the study, you have to wait for
+    // that? Or should it force a separate request? May be slower?
+    // For the moment I just stuck this whole operation in a setTimeout so we are sure the
```

Comment:
Yeah lets not do that. "For the moment" becomes "why the heck was this implemented" 6 months down the line when it causes some esoteric issue.


setTimeout is not ever a valid strategy for assuming something has been grabbed over the network, even if the timeout is many seconds. Maybe your small server hit a bottleneck when you had influx of 100x the uses because of a tweet you made.

Lets just make the datasources cache promises?